core/src/engine/task/task_processor.cpp

```diff
@@ -376,6 +376,35 @@ void TaskProcessor::HandleOverload(
   }
 }
 
+bool TaskProcessor::IsOverloadedByLength(const std::size_t max_queue_length) {
+  bool overloaded_by_length = overloaded_by_length_cache_.load();
+  const int factor = overloaded_by_length ? 4 : 16;
+
+  if (utils::RandRange(factor) != 0) {
+    return overloaded_by_length;
+  }
+
+  overloaded_by_length =
+      ComputeIsOverloadedByLength(overloaded_by_length, max_queue_length);
+  overloaded_by_length_cache_.store(overloaded_by_length,
+                                    std::memory_order_relaxed);
+  return overloaded_by_length;
+}
+
+bool TaskProcessor::ComputeIsOverloadedByLength(
+    const bool current_overloaded_status, const std::size_t max_queue_length) {
+  static constexpr long double kExitOverloadStatusFactor = 0.95;
+  const auto queue_size = GetTaskQueueSize();
+  task_queue_size_overloaded_cache_.store(queue_size,
+                                          std::memory_order_relaxed);
+  if (current_overloaded_status) {
+    return (queue_size >= static_cast<std::size_t>(kExitOverloadStatusFactor *
```

Comment:
Let's micro-optimize here: `max_queue_length - max_queue_length / 16` or something like that